### ISSUE TITLE
Outreachy Documentation Project: Medhavi Singh

### ISSUE THREAD

**Issue by Med16-11** [ersilia-os/ersilia#160]
Applicant: @Med16-11
  
Welcome to the Ersilia Open Source Initiative. This issue will serve to track all your contributions for the project “Improve the documentation and outreach material of the Ersilia Model Hub”.
  
Please tick the tasks as you complete them. To make a final application it is *not* required to have completed all tasks. Only the *Initial Steps* and *Community* sections are REQUIRED. The tasks are not ordered from more to less important, they are simply related to different skills. Start where you feel most comfortable. This project can be adapted to the applicants interests, please focus on the type of tasks that you prefer / have better skills / would like to work on as an intern.

---
### Initial steps:
- [x] Record your application for the project in the outreachy website referencing this issue. Please make sure to select the right project on the website.
- [x] Join the Slack channel to follow public communications
- [x] Comment under this issue explaining why are you interested in this project
---
### GitHub documentation:
- [x] Create a README file with the name <README_yourname.md> under the /documentation folder
- [x] Link the #PR in a comment under this issue
- [x] Incorporate feedback from the mentor
---
### Writing dissemination material
- [x] Read the [Strategic plan 2021-2023](https://ersilia.gitbook.io/ersilia-wiki/) for Ersilia and create a 1-page blogpost with the main points
- [x] Comment under this issue with a link to the blogpost (a google docs for example)
- [ ] Incorporate feedback from the mentor
- [ ] Choose your own topic related to Ersilia (AI/ML for biomedical research, neglected diseases, drug discovery…) and write a 1-page blogpost to communicate to a non-expert audience
- [ ] Comment under this issue with a link to the blogpost (a google docs for example)
- [ ] Incorporate feedback from the mentor
- [ ] Create a template for a twitter post to release every time a new model is incorporated in the Hub (twitter: 280 characters, you can suggest a main post +  thread with extra information) and add it as a comment under this issue
- [ ] Create a template short Newsletter (1 paragraph) to send every month to our community (funders, users, contributors). It should mention metrics (models in the hub, number of users, funding…), thank you etc 
---
### Technical skills (required for the tutorial only)
- [x] [Install](https://ersilia.gitbook.io/ersilia-book/quick-start/installation) the Ersilia Model Hub
- [x] Test one model
- [x] Add a screenshot under this issue showing the model running in your computer
- [ ] Write a docstring for the ErsiliaModel class. Use the Google Python Style guide. Paste the docstring as a comment below (do not use a PR).
---
### Graphic material
- [ ] Read the [Ersilia Brand Guidelines](https://medium.com/ersiliaio/ersilia-open-source-initiative-brand-guidelines-130fe53104f5)
- [ ] Read [“Why Ersilia?”](https://ersilia.gitbook.io/ersilia-wiki/)
- [ ] Create one image / slide to explain Ersilia’s mission and vision
- [ ] Link to the image/slide as a comment under this issue
- [ ] Incorporate feedback from the mentor
- [ ] Create two slides / short video showing how to use the Ersilia Model Hub and add them under the /tutorial folder 
- [ ] Link the #PR in this issue
- [ ] Incorporate feedback from the mentor
---
### Scientific content
- [ ] Check the models available in the [Hub](https://airtable.com/shrXfZ8pqro0jjcsG/tblZGe2a2XeBxrEHP/viwd5XJVLslkE11Tg)
- [ ] Select one model from the list and write a technical card (what is the model for, what input, which data was used to create it, what kind of ML algorithm uses…) for it
- [ ] Add your card as a comment to this issue
- [ ] Search the scientific literature and suggest 3 new models (comment in this issue) that would be relevant to incorporate in the Hub.
---
### Community
- [ ] Look up two other projects and comment on their issues with feedback on one of their tasks
- [ ] If you have feedback from your peers, answer it in this issue.
---
### Other
If you have interest in working on related topics, or have new suggestions, please do the following
- [ ] Add a comment in this issue with your new idea, tagging the mentor 
- [ ] Get feedback from the mentor and act accordingly
- [ ] Link in the comments any other PR you have contributed to.
---
### Final application
- [x] I have answered all comments from mentors and contributors
- [x] All PR or issues assigned to me are complete
- [x] I have submitted my final application to the project

**Comment by Med16-11** [comment]
Hello Mentors,
I am Medhavi Singh, a second-year computer science engineering student.
I like the concept of using artificial intelligence and machine learning to make medicine for infectious diseases. I believe that this type of technology would be very useful for people who are suffering from infectious diseases. Artificial Intelligence and Machine learning have the power to solve these concerning issues. The future of this research work is bright and I want to be part of this revolutionary change. This is the reason why I want to contribute to this project. I can relate to the problems that people suffering from these diseases face. If my contribution can help someone in reducing their pain, I would surely like to contribute.

**Comment by GemmaTuron** [comment]
Hi @Med16-11 

Thanks for your contribution. Please, do not mark tasks before they are completed (I have not yet provided the feedback so wait until I do and you incorporate it, meaning you modify according to my recommendations before marking it as done)
I wanted to ask if perhaps, given your background in computer science, your skills would be well applied to the Code project? Just asking!

**Comment by Med16-11** [comment]
I will keep this thing in mind. 
And yes I would love to contribute to the coding part too.

**Comment by GemmaTuron** [comment]
Hi @Med16-11,

I have merged the readme, I siggest removing the image as it is too detailed, and also is the readme for the whole project so no need for the last section, thanks though
Also please check the licenses it is essential to note which license we use

**Comment by Med16-11** [comment]
Should I make another PR of the suggested changes?

**Comment by Med16-11** [comment]
Hey! @GemmaTuron, I have also installed the Ersilia Model Hub in my system.

![image](https://user-images.githubusercontent.com/78318301/162581595-8e3a0807-073e-4d84-8ba1-fadd85de1785.png)

**Comment by Med16-11** [comment]
Hey @GemmaTuron, I hope you are doing well.
I have completed another task which is writing a page blog post.
Have a look and let me what are the things that needed to be improved.
Google Doc Link: https://docs.google.com/document/d/1Dj8XX69JCBSxjzLBVEmMx6zQsVk78pbxlUUPqsEJQXE/edit?usp=sharing

**Comment by GemmaTuron** [comment]
> Should I make another PR of the suggested changes?

Yes, modify your readme and upload it in a pr, so we have all final documents in one place, thanks

**Comment by GemmaTuron** [comment]
@Med16-11 

Great that the hub is finally running! Perhaps you can focus on one of the technical or scientific tasks to complete your application and submit to the outreachy website

**Comment by Med16-11** [comment]
> > Should I make another PR of the suggested changes?
> 
> Yes, modify your readme and upload it in a pr, so we have all final documents in one place, thanks

> Okay, I will do that. Thanks!

**Comment by Med16-11** [comment]
> @Med16-11
> 
> Great that the hub is finally running! Perhaps you can focus on one of the technical or scientific tasks to complete your application and submit to the outreachy website

> Sure , I will complete the remaining tasks

**Comment by Med16-11** [comment]
from typing import List, Union, Tuple

import numpy as np
from rdkit import Chem
import torch
import torch.nn as nn

from .mpn import MPN
from chemprop.args import TrainArgs
from chemprop.features import BatchMolGraph
from chemprop.nn_utils import get_activation_function, initialize_weights


class MoleculeModel(nn.Module):
    """A :class:`MoleculeModel` is a model which contains a message passing network following by feed-forward layers."""

    def __init__(self, args: TrainArgs, featurizer: bool = False):
        """
	Args:
	    TrainArgs: 
   		A :class:`~chemprop.args.TrainArgs` object containing model arguments.	
            featurizer: 
	    	Whether the model should act as a featurizer, i.e., outputting the
                learned features from the last layer prior to prediction rather than
                outputting the actual property predictions.
        """
        super(MoleculeModel, self).__init__()

        self.classification = args.dataset_type == 'classification'
        self.multiclass = args.dataset_type == 'multiclass'
        self.featurizer = featurizer

        self.output_size = args.num_tasks
        if self.multiclass:
            self.output_size *= args.multiclass_num_classes

        if self.classification:
            self.sigmoid = nn.Sigmoid()

        if self.multiclass:
            self.multiclass_softmax = nn.Softmax(dim=2)

        self.create_encoder(args)
        self.create_ffn(args)

        initialize_weights(self)

    def create_encoder(self, args: TrainArgs) -> None:
        """
        Creates the message passing encoder for the model.
	
	Args:
            TrainArgs: 
		A :class:`~chemprop.args.TrainArgs` object containing model arguments.

        """
        self.encoder = MPN(args)
              
        if args.checkpoint_frzn is not None:
            if args.freeze_first_only: # Freeze only the first encoder
                for param in list(self.encoder.encoder.children())[0].parameters():
                    param.requires_grad=False
            else: # Freeze all encoders
                for param in self.encoder.parameters():
                    param.requires_grad=False                   
                        
    def create_ffn(self, args: TrainArgs) -> None:
        """
        Creates the feed-forward layers for the model.
	
	Args:
            TrainArgs: 
                A :class:`~chemprop.args.TrainArgs` object containing model arguments.

        """
        self.multiclass = args.dataset_type == 'multiclass'
        if self.multiclass:
            self.num_classes = args.multiclass_num_classes
        if args.features_only:
            first_linear_dim = args.features_size
        else:
            first_linear_dim = args.hidden_size * args.number_of_molecules
            if args.use_input_features:
                first_linear_dim += args.features_size

        if args.atom_descriptors == 'descriptor':
            first_linear_dim += args.atom_descriptors_size

        dropout = nn.Dropout(args.dropout)
        activation = get_activation_function(args.activation)

        # Create FFN layers
        if args.ffn_num_layers == 1:
            ffn = [
                dropout,
                nn.Linear(first_linear_dim, self.output_size)
            ]
        else:
            ffn = [
                dropout,
                nn.Linear(first_linear_dim, args.ffn_hidden_size)
            ]
            for _ in range(args.ffn_num_layers - 2):
                ffn.extend([
                    activation,
                    dropout,
                    nn.Linear(args.ffn_hidden_size, args.ffn_hidden_size),
                ])
            ffn.extend([
                activation,
                dropout,
                nn.Linear(args.ffn_hidden_size, self.output_size),
            ])
            

        # Create FFN model
        self.ffn = nn.Sequential(*ffn)
        
        if args.checkpoint_frzn is not None:
            if args.frzn_ffn_layers >0:
                for param in list(self.ffn.parameters())[0:2*args.frzn_ffn_layers]: # Freeze weights and bias for given number of layers
                    param.requires_grad=False


    def featurize(self,
                  batch: Union[List[List[str]], List[List[Chem.Mol]], List[List[Tuple[Chem.Mol, Chem.Mol]]], List[BatchMolGraph]],
                  features_batch: List[np.ndarray] = None,
                  atom_descriptors_batch: List[np.ndarray] = None,
                  atom_features_batch: List[np.ndarray] = None,
                  bond_features_batch: List[np.ndarray] = None) -> torch.FloatTensor:
        """
        Computes feature vectors of the input by running the model except for the last layer.

	Args:
          batch: 
	    A list of list of SMILES, a list of list of RDKit molecules, or a
            list of :class:`~chemprop.features.featurization.BatchMolGraph`.
            The outer list or BatchMolGraph is of length :code:`num_molecules` (number of datapoints in batch),
            the inner list is of length :code:`number_of_molecules` (number of molecules per datapoint).
          features_batch: 
	    A list of numpy arrays containing additional features.
          atom_descriptors_batch: 
            A list of numpy arrays containing additional atom descriptors.
          atom_features_batch: 
            A list of numpy arrays containing additional atom features.
          bond_features_batch: 
            A list of numpy arrays containing additional bond features.
        
	Returns: 
		The feature vectors computed by the :class:`MoleculeModel`.
        """
        return self.ffn[:-1](self.encoder(batch, features_batch, atom_descriptors_batch,
                                          atom_features_batch, bond_features_batch))

    def fingerprint(self,
                  batch: Union[List[List[str]], List[List[Chem.Mol]], List[List[Tuple[Chem.Mol, Chem.Mol]]], List[BatchMolGraph]],
                  features_batch: List[np.ndarray] = None,
                  atom_descriptors_batch: List[np.ndarray] = None) -> torch.FloatTensor:
        """
        Encodes the fingerprint vectors of the input molecules by passing the inputs through the MPNN and returning
        the latent representation before the FFNN.
	
	Args:
         batch: 
	     A list of list of SMILES, a list of list of RDKit molecules, or a
             list of :class:`~chemprop.features.featurization.BatchMolGraph`.
             The outer list or BatchMolGraph is of length :code:`num_molecules` (number of datapoints in batch),
             the inner list is of length :code:`number_of_molecules` (number of molecules per datapoint).
         features_batch: 
	     A list of numpy arrays containing additional features.
         atom_descriptors_batch: 
             A list of numpy arrays containing additional atom descriptors.

        Returns: 
	     The fingerprint vectors calculated through the MPNN.

        """
        return self.encoder(batch, features_batch, atom_descriptors_batch)

    def forward(self,
                batch: Union[List[List[str]], List[List[Chem.Mol]], List[List[Tuple[Chem.Mol, Chem.Mol]]], List[BatchMolGraph]],
                features_batch: List[np.ndarray] = None,
                atom_descriptors_batch: List[np.ndarray] = None,
                atom_features_batch: List[np.ndarray] = None,
                bond_features_batch: List[np.ndarray] = None) -> torch.FloatTensor:
        """
        Runs the :class:`MoleculeModel` on input.
	
	Args:
            batch: 
                A list of list of SMILES, a list of list of RDKit molecules, or a
                list of :class:`~chemprop.features.featurization.BatchMolGraph`.
                The outer list or BatchMolGraph is of length :code:`num_molecules` (number of datapoints in batch),
                the inner list is of length :code:`number_of_molecules` (number of molecules per datapoint).
            features_batch: 
                A list of numpy arrays containing additional features.
            atom_descriptors_batch: 
                A list of numpy arrays containing additional atom descriptors.
            atom_features_batch: 
                A list of numpy arrays containing additional atom features.
            bond_features_batch: 
                A list of numpy arrays containing additional bond features.

        Returns: 
	     The output of the :class:`MoleculeModel`, which is either property predictions
             or molecule features if :code:`self.featurizer=True`.
        """
        if self.featurizer:
            return self.featurize(batch, features_batch, atom_descriptors_batch,
                                  atom_features_batch, bond_features_batch)

        output = self.ffn(self.encoder(batch, features_batch, atom_descriptors_batch,
                                       atom_features_batch, bond_features_batch))

        # Don't apply sigmoid during training b/c using BCEWithLogitsLoss
        if self.classification and not self.training:
            output = self.sigmoid(output)
        if self.multiclass:
            output = output.reshape((output.size(0), -1, self.num_classes))  # batch size x num targets x num classes per target
            if not self.training:
                output = self.multiclass_softmax(output)  # to get probabilities during evaluation, but not during training as we're using CrossEntropyLoss

        return output
       


I have tried writing docstring for this piece of code according to the google python style guide. I am not sure why the formatting seems a bit disoriented. @GemmaTuron Do let me know, Am I going in right direction or need to make changes ?
Reference: https://github.com/ersilia-os/eos4e40/blob/main/model/framework/chemprop/models/model.py

Thank you!

**Comment by GemmaTuron** [comment]
Hi @Med16-11 


I think you have chosen quite a complex Module to document. My suggestions are:
- if you do not have expertise in python documentation writing, its fine it is not a requirement for the project so don't focus on this task
- If you still want to try it out choose a simpler module from the ersilia repository, not the eos4e40.

**Comment by Med16-11** [comment]
Okay! I will complete my Outreachy applications with the help of other tasks that I did. 
I will continue doing these tasks since there are a lot of things in which I am interested and want to contribute. One of them is Graphic material tasks. Thanks for the suggestions.
Also, I have already raised a PR that contains all the changes that you suggested, Please have a look.

Thank you!

**Comment by GemmaTuron** [comment]
Hi @Med16-11 

Thanks, I have merged your PR.
Please focus on the final application first before thinking of any new tasks, since the deadline is approaching.

**Comment by Med16-11** [comment]
Thanks for merging the PR.
Also I have submitted my final application.
